File: backend/chess4connect/chess4connect/Models/SocketComunication/MessageTypes/SocketMessage.cs
using chess4connect.Enums;

namespace chess4connect.Models.SocketComunication.MessageTypes;

public class SocketMessage
{
    public virtual SocketCommunicationType Type { get; set; }
}

public class SocketMessage<T> : SocketMessage
{
    public T Data { get; set; }
}


public class GameSocketMessage<T> : SocketMessage
{
    public override SocketCommunicationType Type => SocketCommunicationType.GAME;
    public T Data { get; set; }
}
public class ConnectionSocketMessage<T> : SocketMessage
{
    public override SocketCommunicationType Type => SocketCommunicationType.CONNECTION;
    public T Data { get; set; }
}

public class ChatSocketMessage<T> : SocketMessage
{
    public override SocketCommunicationType Type => SocketCommunicationType.CHAT;
    public T Data { get; set; }
}

File: frontend/chessandconnect/src/app/models/WebSocketMessages/SocketMessage.ts
import { SocketCommunicationType } from "../../enums/SocketCommunicationType";

export class SocketMessage {
    Type!: SocketCommunicationType;
}

export class SocketMessageGeneric<T> extends SocketMessage {
    Data!: T;
}

export class GameSocketMessage<T> extends SocketMessageGeneric<T> {
    override Type: SocketCommunicationType = SocketCommunicationType.GAME;
}

export class ConnectionSocketMessage<T> extends SocketMessageGeneric<T> {
    override Type: SocketCommunicationType = SocketCommunicationType.CONNECTION;
}

export class ChatSocketMessage<T> extends SocketMessageGeneric<T> {
    override Type: SocketCommunicationType = SocketCommunicationType.CHAT;
}
